--- kobs-ng-3.10.17-1.0.0.ori/src/mtd.h	2014-01-16 22:38:15.000000000 +0800
+++ kobs-ng-3.10.17-1.0.0/src/mtd.h	2014-07-14 13:50:27.609109953 +0800
@@ -31,6 +31,9 @@
 #include "BootControlBlocks.h"
 #include "rom_nand_hamming_code_ecc.h"
 
+#define MTD_FILE_MODE_NORMAL 0
+#define MTD_FILE_MODE_RAW 3
+
 //------------------------------------------------------------------------------
 // Re-definitions of true and false, because the standard ones aren't good
 // enough?
